glyphs2fontir/src/source.rs

```diff
@@ -258,20 +276,97 @@ impl Work for StaticMetadataWork {
+        .map(|(axis, pos)| (axis.tag.clone(), DesignCoord::new(pos.into_inner() as f32)))
+        .collect()
+}
+
+fn to_normalized(axes: &[ir::Axis], design_location: &DesignLocation) -> NormalizedLocation {
+    let axes: HashMap<_, _> = axes.iter().map(|a| (&a.tag, a)).collect();
+    design_location
+        .iter()
+        .map(|(tag, dc)| {
+            (
+                tag.clone(),
+                dc.to_normalized(&axes.get(tag).unwrap().converter),
+            )
+        })
+        .collect()
+}
+
 struct GlyphIrWork {
     glyph_name: String,
+    font: Arc<Font>,
+    master_indices: Arc<HashMap<String, usize>>,
 }
 
 impl Work for GlyphIrWork {
     fn exec(&self, context: &Context) -> Result<(), WorkError> {
         debug!("Generate IR for {}", self.glyph_name);
+        let font = self.font.as_ref();
+
         let static_metadata = context.get_static_metadata();
+        let axes = &static_metadata.axes;
+
         let gid = static_metadata
             .glyph_id(&self.glyph_name)
             .ok_or_else(|| WorkError::NoGlyphIdForName(self.glyph_name.clone()))?;
+        let glyph = font
+            .glyphs
+            .get(&self.glyph_name)
+            .ok_or_else(|| WorkError::NoGlyphForName(self.glyph_name.clone()))?;
+
+        let mut sources = HashMap::new();
+
+        // Glyphs have layers that match up with masters, and masters have locations
+        let mut axis_positions: HashMap<String, HashSet<NormalizedCoord>> = HashMap::new();
+        for layer in glyph.layers.iter() {
+            let Some(master_idx) = self.master_indices.get(&layer.layer_id) else {
+                panic!("glyph {} references layer id {} but no such master exists", self.glyph_name, layer.layer_id);
+            };
+            let master = &font.font_master[*master_idx];
+            let location = to_normalized(axes, &design_location(axes, master));
+
+            for (tag, coord) in location.iter() {
+                axis_positions
+                    .entry(tag.clone())
+                    .or_default()
+                    .insert(*coord);
+            }
+
+            // TODO actually populate fields
+            let glyph_instance = GlyphInstance {
+                width: 0_f64,
+                height: None,
+                contours: Vec::new(),
+                components: Vec::new(),
+            };
+
+            sources.insert(location, glyph_instance);
+        }
+
+        // It's helpful if glyphs are defined at min, default, and max (some of which may be cooincident)
+        for axis in axes.iter() {
+            let min = axis.min.to_normalized(&axis.converter);
+            let max = axis.max.to_normalized(&axis.converter);
+            let default = axis.max.to_normalized(&axis.converter);
+            let positions = axis_positions.get(&axis.tag).unwrap();
+            if !positions.contains(&min) {
+                panic!("{} is undefined at {} min", self.glyph_name, axis.tag);
```

Comment:
I think panic! should be used sparingly.. basically a bit like AssertionError in Python, it must never happen in normal nor in exceptional but still predictable circumstances where you would signal with an Exception in Python or Result::Err in Rust. If this can never happen and it's about inner code logic, then it's fine to keep; otherwise, if it can happen sometimes that such condition isn't met then an appropriate WorkError should be returned, I suppose

Reply:
Fair point. I was actually just thinking we have a bit more unwrap/panic than I'd like.

Reply:
Made it an error

Reply:
did you? I still see panic! here

Reply:
Err. I swear I did. Did I not upload it?!

Reply:
Fixed in #63